File: app/src/main/java/com/wsoteam/diet/presentation/activity/FavoriteSource.java
package com.wsoteam.diet.presentation.activity;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.NonNull;
import io.reactivex.Flowable;
import io.reactivex.Single;
import java.util.List;

public class FavoriteSource extends ExercisesSource {

  private SharedPreferences preferences;
  private AssetsSource origin;

  public FavoriteSource(Context context) {
    this.preferences = context.getSharedPreferences("favorite_exercises", Context.MODE_PRIVATE);
    this.origin = new AssetsSource(context.getAssets());
  }

  @Override public Single<List<UserActivityExercise>> all() {
    return origin.all()
        .flatMapPublisher(Flowable::fromIterable)
        .filter(e -> preferences.getBoolean(e.getTitle().toString(), false))
        .toList();
  }

  @Override public Single<UserActivityExercise> add(@NonNull UserActivityExercise exercise) {
    preferences.edit()
        .putBoolean(exercise.getTitle().toString(), true)
        .apply();

    return Single.just(exercise);
  }

  @Override public Single<UserActivityExercise> edit(@NonNull UserActivityExercise exercise) {
    throw new UnsupportedOperationException("can't edit favorite items");
  }

  @Override public Single<UserActivityExercise> remove(@NonNull UserActivityExercise exercise) {
    preferences.edit()
        .remove(exercise.getTitle().toString())
        .apply();

    return Single.just(exercise);
  }
}

File: app/src/main/java/com/wsoteam/diet/presentation/activity/FavoriteSource.kt
package com.wsoteam.diet.presentation.activity

import com.wsoteam.diet.presentation.activity.ActivitiesSyncedSource.Companion.ActivitySource.CUSTOM
import io.reactivex.Single

class FavoriteSource : ActivitiesSyncedSource(CUSTOM) {
  override var filterFavorites = true
  override var filterByDate: Boolean = false

  override fun add(exercise: ActivityModel): Single<ActivityModel> {
    exercise as UserActivityExercise

    return super.add(exercise.copy(favorite = true, `when` = System.currentTimeMillis()))
  }
}
